rdmo/projects/viewsets.py

```diff
@@ -521,29 +528,33 @@ def dispatch(self, *args, **kwargs):
 
     def retrieve(self, request, *args, **kwargs):
         page = self.get_object()
-        conditions = page.conditions.select_related('source', 'target_option')
-
+        catalog = self.project.catalog
         values = self.project.values.filter(snapshot=None).select_related('attribute', 'option')
 
-        if check_conditions(conditions, values):
+        sets = compute_sets(values)
+        resolved_conditions = resolve_conditions(catalog, values, sets)
+
+        # check if the current page meets conditions
+        if compute_show_page(page, resolved_conditions):
             serializer = self.get_serializer(page)
             return Response(serializer.data)
         else:
-            if request.GET.get('back') == 'true':
-                prev_page = self.project.catalog.get_prev_page(page)
-                if prev_page is not None:
-                    url = reverse('v1-projects:project-page-detail',
-                                  args=[self.project.id, prev_page.id]) + '?back=true'
-                    return HttpResponseRedirect(url, status=303)
-            else:
-                next_page = self.project.catalog.get_next_page(page)
-                if next_page is not None:
-                    url = reverse('v1-projects:project-page-detail', args=[self.project.id, next_page.id])
-                    return HttpResponseRedirect(url, status=303)
-
-            # indicate end of catalog
+            # determine the direction of navigation (previous or next)
+            direction = 'prev' if request.GET.get('back') == 'true' else 'next'
+
+            # find the next relevant page with from pages and resolved conditions
+            next_page = compute_next_relevant_page(page, direction, catalog, resolved_conditions)
+
+            if next_page is not None:
+                url = reverse('v1-projects:project-page-detail', args=[self.project.id, next_page.id])
+                if direction == 'prev':
```

Comment:
I think we don't need `back` here anymore, it was only used for the conditions check.